### PR TITLE
Suspend the scheduler when starting a root sagas

### LINKED ISSUE #1623: Suspend before running root
root should be considered a regular task (its just being run from outside instead of being forked inside) and thus normal scheduling should apply for it.

So we should suspend the scheduler before running, I can dig up a test case for this as I had it somewhere

### PR DESCRIPTION
**Pull request by Andarist**
fixes #1623

### CODE REVIEW

File: packages/core/__tests__/interpreter/takeSync.js [hunk 1]
```diff
@@ -643,3 +654,29 @@ test('put causing sync dispatch response in store subscriber', () => {
...
+    return state.concat(action.type)
+  }
+
+  const middleware = sagaMiddleware()
+  const store = createStore(reducer, applyMiddleware(middleware))
+
+  function* root() {
+    yield all([
+      put({ type: 'FIRST' }),
+      takeEvery('FIRST', function*() {
```
Comment (restrry) [inline packages/core/__tests__/interpreter/takeSync.js]: what about such case, where we use use `proc` internally for iterable
```js
function* root() {
    yield all([
      put({ type: 'FIRST' }),
      function*() {
        yield take('FIRST')
        yield put({ type: 'SECOND' })
      },
    ])
  }

```

Comment (Andarist) [reply packages/core/__tests__/interpreter/takeSync.js]: aren't they the same? `takeEvery` is also using proc internally, as it's yielding `fork` which is wrapped with proc.

File: packages/core/src/internal/runSaga.js [hunk 2]
```diff
@@ -85,11 +86,14 @@ export function runSaga(options, saga, ...args) {
     finalizeRunEffect,
   }
 
+  suspend()
   const task = proc(env, iterator, context, effectId, getMetaInfo(saga), null)
```
Comment (shinima) [inline packages/core/src/internal/runSaga.js]: Could we move `suspend()` and `flush()` into `proc` as follows?

```javascript
export default function proc(env, iterator, parentContext, parentEffectId, meta, cont) {
  suspend()
  // other code...
  flush()
  return task
```

If we do this, we just need write `suspend()` and `flush()` **one time**. Not sure this idea would work, but tests are passing in my PC.

Comment (shinima) [reply packages/core/src/internal/runSaga.js]: Further more, if we do like this, we may export a function `wrap` (or some other better name) from *packages/core/src/internal/scheduler.js* and use it as a decorator on `proc()`.

```javascript
export function wrap(fn) {
  return (...args) => {
    try {
      suspend()
      return fn(...args)
    } finally {
      flush()
    }
  }
}
```

File: packages/core/__tests__/interpreter/takeSync.js (hunk 1 above)
Comment (restrry) [reply packages/core/__tests__/interpreter/takeSync.js]: there is no `takeEvery`, if you add this test you will see that it fails

Comment (Andarist) [reply packages/core/__tests__/interpreter/takeSync.js]: good catch! should be handled as well

File: packages/core/src/internal/runSaga.js (hunk 2 above)
Comment (Andarist) [reply packages/core/src/internal/runSaga.js]: would prefer avoiding additional closures in the saga code, so Im reluctant to using decorator approach

and Im not sure if we can move this safely into the proc, as `runForkEffect` would *slightly* change the order of how things happen (when the cb callback gets invoked) - I will experiment with it though

Comment (Andarist) [reply packages/core/src/internal/runSaga.js]: It seems that tests are passing if we move this to proc, I'll try to do further experimentation to see if I can break some case

File: packages/core/__tests__/interpreter/takeSync.js (hunk 1 above)
Comment (Andarist) [reply packages/core/__tests__/interpreter/takeSync.js]: actually ur given snippet doesnt work, as it yields a generator and not the iterator, even if we correct it to:
```js
function* root() {
    yield all([
      put({ type: 'FIRST' }),
      (function*() {
        yield take('FIRST')
        yield put({ type: 'SECOND' })
      })(),
    ])
  }
```

the second yielded item in all is equivalent to `call` effect (kinda) and thus scheduler doesn't apply to it

or did I miss something?

**Comment by Andarist** [comment]
I've added a test and wrapped root saga's proc in try/finally - in the same way runForkEffect is already, to keep the scheduler's state more safe.

@shinima can't move suspend+flush into proc, because proc is also used for call effects which are not scheduler and because flush has to happen in runForkEffect AFTER the the task descriptor gets returned to the caller and not before that

File: packages/core/__tests__/interpreter/takeSync.js [hunk 3]
```diff
@@ -643,3 +654,54 @@ test('put causing sync dispatch response in store subscriber', () => {
...
+  const actual = []
+  const reducer = (state, action) => action.type
+
+  const middleware = sagaMiddleware()
+  const store = createStore(reducer, applyMiddleware(middleware))
+
+  function* root() {
+    // force async, otherwise sync root startup prevents this from being tested appropriately
+    // as the scheduler is in suspended state because of it
+    yield delay(10)
```
Comment (shinima) [inline packages/core/__tests__/interpreter/takeSync.js]: We can replace it with `yield Promise.resolve()` for faster execution.

**Pull request merged**